### PR TITLE
miagrate pr [LLM Runtime] Add Whisper Example and Python API

### PR DESCRIPTION
**Pull request by intellinjun**
## Type of Change

feature or bug fix or documentation or others
API changed or not

## Description

detail description 
Issues: xxx

## Expected Behavior & Potential Risk

the expected behavior that triggered by this PR 

## How has this PR been tested?

how to reproduce the test (including hardware information)

## Dependency Change?

any library dependency introduced or removed

### CODE REVIEW

**Review by a32543254 — APPROVED** [review]
LGTM

**Comment by intellinjun** [comment]
https://inteltf-jenk.sh.intel.com/job/neural_speed_extension/17/

**Comment by airMeng** [comment]
consider uncommenting this if you are free
https://github.com/intel/neural-speed/blob/749caca6e94b88a8e105f9f4f6ab036acf2d87a0/.clang-tidy#L3

File: scripts/convert.py [hunk 1]
```diff
@@ -26,6 +26,7 @@ def main(args_in: Optional[List[str]] = None) -> None:
         default="f32",
     )
     parser.add_argument("--outfile", type=Path, required=True, help="path to write to")
+    parser.add_argument("--whisper_repo_path", type=Path, required=False, help="path to whisper repo")
```
Comment (airMeng) [inline scripts/convert.py:29]: what kinds of path, can you clarify on readme?

File: neural_speed/models/model_utils/quant_utils.h [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+#define MODEL_FILE_MAGIC_GGMF 0x67676d66u  // 'ggmf'
+#define MODEL_FILE_MAGIC_NE 0x67676d6cu    // 'ne'
+#define MODEL_FILE_MAGIC_GGSN 0x6767736eu  // 'ggsn'
+
+#define MODEL_FILE_VERSION 3
+#define MODEL_FILE_MAGIC MODEL_FILE_MAGIC_GGJT
+#define MODEL_FILE_MAGIC_UNVERSIONED MODEL_FILE_MAGIC_NE
+#define MODEL_SESSION_MAGIC MODEL_FILE_MAGIC_GGSN
+#define MODEL_SESSION_VERSION 1
+
```
Comment (airMeng) [inline neural_speed/models/model_utils/quant_utils.h:45]: check whether these are useful in this file

File: scripts/convert.py (hunk 1 above)
Comment (intellinjun) [reply scripts/convert.py:29]: I will rm it.

File: neural_speed/application/common.cpp [hunk 3]
```diff
@@ -102,6 +102,8 @@ bool isValidFilename(const std::string& filename) {
   return infile.good();
 }
 
+int64_t common_time_us() { return ne_time_us(); }
```
[the remark this answers is not in the document]
Comment (intellinjun) [reply neural_speed/application/common.cpp:105]: you are right, I will replace it.

**Comment by airMeng** [comment]
> > consider uncommenting this if you are free
> > https://github.com/intel/neural-speed/blob/749caca6e94b88a8e105f9f4f6ab036acf2d87a0/.clang-tidy#L3
> 
> I fix some format error, but some error ，however, some variables need to be modified in sync with bestla.

Ok for me,  you can list the bestla issues here

**Comment by intellinjun** [comment]
> > > consider uncommenting this if you are free
> > > https://github.com/intel/neural-speed/blob/749caca6e94b88a8e105f9f4f6ab036acf2d87a0/.clang-tidy#L3
> > 
> > 
> > I fix some format error, but some error ，however, some variables need to be modified in sync with bestla.
> 
> Ok for me, you can list the bestla issues here
Can you please fix this error? This is related to bestla，there is some variables in bestla
/neural-speed/neural_speed/core/layers/mha_dense.cpp:315:10: warning: invalid case style for struct 'Param' [readability-identifier-naming]
   struct Param {
       ^~~~~
         param_t

File: neural_speed/models/whisper/whisper.cpp [hunk 4]
```diff
@@ -2434,9 +2433,9 @@ void whisper_free(struct whisper_context* ctx) {
     if (ctx->model.ctx) {
       ne_free(ctx->model.ctx);
     }
-    if (ctx->model.buf) {
-      delete ctx->model.buf;
-    }
+    // if (ctx->model.buf) {
+    //   delete ctx->model.buf;
+    // }
```
Comment (airMeng) [inline neural_speed/models/whisper/whisper.cpp:2438]: why not delete?

File: neural_speed/convert/convert_whisper.py [hunk 5]
```diff
@@ -88,138 +91,148 @@ def bytes_to_unicode():
...
+            # data
+            data.tofile(fout)
+
+        fout.close()
+
+        print("Done. Output file: ", fname_out)
+        print("")
+    
+    if __name__ == "__main__":
+        main()
```
Comment (airMeng) [inline neural_speed/convert/convert_whisper.py:238]: no new line


you can set it in vscode settings

File: neural_speed/application/whisper_pybind.cpp [hunk 6]
```diff
@@ -0,0 +1,472 @@
...
+  m.doc() = "cpp model python binding";
+  py::class_<Model>(m, "Model", py::module_local())
+      .def(py::init())
+      .def("init_model", &Model::init_model, "initial model with model path and parameters", py::arg("model_path"))
+      .def_static("quant_model", &Model::quant_model, "Quantize model", py::arg("model_path"), py::arg("out_path"),
+                  py::arg("weight_dtype") = "int4", py::arg("alg") = "sym", py::arg("group_size") = 32,
+                  py::arg("scale_dtype") = "fp32", py::arg("compute_dtype") = "int8", py::arg("use_ggml") = false,
+                  py::arg("threads") = 8)
+      .def("inference", &Model::inference, "Translate audio to text");
+}
```
Comment (airMeng) [inline neural_speed/application/whisper_pybind.cpp:472]: / no new line

**Review by airMeng — APPROVED** [review]

**Review by zhenwei-intel — APPROVED** [review]

**Comment by intellinjun** [comment]
https://inteltf-jenk.sh.intel.com/job/neural_speed_extension/29/artifact/

**Comment by intellinjun** [comment]
https://[inteltf-jenk.sh.intel.com/blue/rest/organizations/jenkins/pipelines/neural_speed_cpp_graph_test/runs/215/nodes/77/steps/80/log/?start=0](https://inteltf-jenk.sh.intel.com/blue/rest/organizations/jenkins/pipelines/neural_speed_cpp_graph_test/runs/215/nodes/77/steps/80/log/?start=0)

**Pull request merged**